### DISCUSSION TITLE
Miller 5 and 6 and managing a JSON

### DISCUSSION THREAD

**Discussion by aborruso** [johnkerl/miller#1667] [Q&A]
Dear @johnkerl ,
in an old miller 5 script of mine I run

```
curl -kL 'https://gist.githubusercontent.com/aborruso/c0599a96c816fa57fa26912b38948b16/raw/0f195d860b18fdb14528935e76a7625ac998fd3b/tmp.json' | mlr5 --j2c cat then cut -x -r -f ":"
```

I have no error. In Miller 6 there are a lot of new things related to the Json world, but I can't figure out how to get the same result that 5 gave me.

The complete mlr5 code is this one

```
curl -kL 'https://gist.githubusercontent.com/aborruso/c0599a96c816fa57fa26912b38948b16/raw/0f195d860b18fdb14528935e76a7625ac998fd3b/tmp.json' | mlr5 --j2c cut -x -r -f ":" then unsparsify then sort -r metadata_created then put '$URL="https://dati.gov.it/view-dataset/dataset?id=".$id' then reorder -f metadata_created,metadata_modified,title,holder_name,url,URL,holder_identifier,issued,modified then put 'if(is_null($url)){$url=$indirizzo}' 
```

And I obtain something similar to the below table.

If I use miller 6

```
curl -kL 'https://gist.githubusercontent.com/aborruso/c0599a96c816fa57fa26912b38948b16/raw/0f195d860b18fdb14528935e76a7625ac998fd3b/tmp.json' | mlr --flatsep : --j2c cat then cut -x -r -f ":"
```

I have `mlr: exiting due to data error.`  I tried using `flatten` and `unflatten` but I always get errors.

How to obtain in mlr6 the same output I have using mlr5?

Thank you

| metadata_created | metadata_modified | title | holder_name | url | URL | holder_identifier | issued | modified | author | author_email | creator_user_id | dataset_is_local | id | isopen | license_id | license_title | maintainer | maintainer_email | name | notes | num_resources | num_tags | owner_org | private | state | type | version | indirizzo | license_url |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 2024-09-19T12:03:03.238952 | 2024-09-19T12:03:03.238957 | Interventi pnrr-pnc | Comune di Torino | Https://www.torinocambia.it/ | https://dati.gov.it/view-dataset/dataset?id=7570f7ff-61e8-427f-88e8-2e6e3354385c | 00514490010 | 18-04-2024 | 18-04-2024 |  |  | e514c9c4-6d64-45d6-ae28-c70bd3592d59 | false | 7570f7ff-61e8-427f-88e8-2e6e3354385c | true | cc-by-sa | Creative Commons Attribution Share-Alike |  |  | interventi-pnrr-pnc-0424 | Riepilogo degli interventi pnrr e pnc | 1 | 1 | f804b34c-a5f7-42ca-bde2-82a0bcadd55e | false | active | dataset | 1.0 |  | http://www.opendefinition.org/licenses/cc-by-sa |

**Comment by aborruso** [comment]
@johnkerl I'm tagging you, just to make sure you get the notification. Sorry for the double tagging.

**Comment by KeelMatrix** [comment]
The main change is that `--j2c` is gone: Miller 6 flattens JSON on ingest, so JSON→CSV is just the input/output flag pair.

```
mlr5 --j2c ...        →        mlr6 --ijson --ocsv ...
```

The second change is the one that bites. Miller 6 flattens nested JSON with `.` (arrays get 1-based positions: `b.d.1`, `b.d.2`), where Miller 5 used `:`. So `cut -x -r -f ":"` no longer drops anything — the nested names don't contain a colon any more — and on a CKAN dump where records don't all carry the same keys, letting the CSV writer see those nested fields gives you `mlr: CSV schema change` and a truncated file.

Selecting the columns you actually want fixes both, and it's what makes the header stable. This is your gist, unchanged, run through Miller 6.21:

```console
$ curl -kL 'https://gist.githubusercontent.com/aborruso/c0599a96c816fa57fa26912b38948b16/raw/0f195d860b18fdb14528935e76a7625ac998fd3b/tmp.json' \
  | mlr --ijson --ocsv \
      cut -f metadata_created,metadata_modified,title,holder_name,url,holder_identifier,issued,modified,indirizzo \
      then unsparsify \
      then put 'if (is_null($url)) { $url = $indirizzo }' \
      then reorder -f metadata_created,metadata_modified,title,holder_name,url,holder_identifier,issued,modified \
      then put '$URL = "https://dati.gov.it/view-dataset/dataset?id=" . $id'
```

```console
metadata_created,metadata_modified,title,holder_name,url,holder_identifier,issued,modified,indirizzo,URL
2024-05-07T16:20:51.421240,2024-09-28T07:25:15.676951,Regione Toscana - PNRR,Regione Toscana,https://dati.toscana.it/dataset/7083567c-92e3-4d17-83c4-438f62d78bf6,r_toscan,2022-07-19,2024-03-20,https://dati.toscana.it/dataset/7083567c-92e3-4d17-83c4-438f62d78bf6,https://dati.gov.it/view-dataset/dataset?id=7570f7ff-61e8-427f-88e8-2e6e3354385c
2024-09-19T12:03:03.238952,2024-09-19T12:03:03.238957,Interventi pnrr-pnc,Comune di Torino,Https://www.torinocambia.it/,00514490010,18-04-2024,18-04-2024,,https://dati.gov.it/view-dataset/dataset?id=7570f7ff-61e8-427f-88e8-2e6e3354385c
```

Three details that cost me some time:

- `unsparsify` has to run before the CSV writer for ragged records. `cut -f <cols> then unsparsify` is the combination that ends up with one consistent header; if you keep all top-level fields instead of selecting them, the key sets differ per record and the writer errors out with the schema-change message.
- Your `is_null($url)` fallback still works in Miller 6, but the reason is worth knowing: `is_null` is true for JSON null *and* for empty strings (`{"a":""}` → `is_null($a)` is `true`), while `is_empty` is false for a JSON null. So don't swap `is_null` for `is_empty` thinking they're equivalent — on your Toscana record (`"url": ""`, `indirizzo` set) that swap is what keeps the URL column populated.
- If you want only the top-level scalars of an arbitrary record, Miller 6 can drop a nested subtree by name: `cut -x -f b` removes `b.c`, `b.d.1`, `b.d.2`, ... That's the closest equivalent to your Miller 5 `:` filter, because the regex route no longer matches flattened JSON field names.

Disclosure: I work on jsonnorm.com, an open JSON→CSV converter. The nested-array fork you're working around (keep indexed columns vs one row per element) is the part most converters leave implicit; this is that same decision made explicit, not a replacement for `mlr`: https://jsonnorm.com/convert/nested-json-to-csv/?src=ghdisc
